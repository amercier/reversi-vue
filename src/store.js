import Vue from 'vue';
import Vuex from 'vuex';
import { Game, PIECE_TYPES } from 'reversi';
import { BOARD_SIZE } from './config';

Vue.use(Vuex);

/**
 * Core logics for the Reversi.
 * @namespace reversi
 * @see NPM package documentation: https://www.npmjs.com/package/reversi
 *
 * @typedef {Object} Game
 * @memberof reversi
 *
 * @typedef {Object} Board
 * @memberof reversi
 *
 * @typedef {Object} Square
 * @memberof reversi
 *
 * @var PIECE_TYPES
 * @property {string} BLANK Empty cell.
 * @property {string} WHITE White stone.
 * @property {string} BLACK Black stone.
 * @memberof reversi
 */

/**
 * @typedef {Object} Cell
 * @property {number} row Row index.
 * @property {number} col Column index.
 * @property {number} isStone Row index.
 * @property {boolean} isStone Whether the cell contains a stone.
 * @property {boolean} isMove Whether the cell contains a potential move.
 * @property {string} color Color for the stone or move.
 */

/**
 * Get a square-to-cell converter.
 *
 * @example
 *   const { board } = new Game();
 *   const squares = board.squares[0]; // First row
 *   const cells = squares.map(getSquareToCell(board, PIECE_TYPES.BLACK);
 *
 * @param {Board} board - A game board.
 * @param {string} nextPieceType - Next color to play: either PIECE_TYPES.WHITE or PIECE_TYPES.BLACK.
 * @returns {squareToCell} The square-to-cell converter.
 *
 * @see Reversi package: https://www.npmjs.com/package/reversi
 */
export function getSquareToCell(board, nextPieceType) {
  /**
   * @typedef {Function} squareToCell
   * @example
   *   squareToCell({ rowIndex: 0, colIndex: 0, pieceType: 'BLACK' })
   * @param {Square} square - A reversi square.
   * @returns {Cell} The cell corresponding to `square`.
   */
  return function squareToCell({ rowIndex: row, colIndex: col, pieceType }) {
    const isMove =
      pieceType === PIECE_TYPES.BLANK &&
      board.isPlaceableSquare(row, col, nextPieceType);
    return {
      row,
      col,
      isStone: pieceType !== PIECE_TYPES.BLANK,
      isMove,
      color: isMove
        ? board.isPlaceableSquare(row, col, nextPieceType)
          ? nextPieceType.toLowerCase()
          : null
        : pieceType.toLowerCase(),
    };
  };
}

/**
 * Create a new store.
 *
 * @example
 *   createStore();
 * @returns {Vuex.Store} A new store.
 */
export const createStore = () =>
  new Vuex.Store({
    /**
     * Store state
     *
     * @type {Object}
     * @see https://vuex.vuejs.org/guide/state.html
     */
    state: {
      /**
       * The `reversi` game instance.
       *
       * @type {Game}
       */
      game: new Game(),

      /**
       * Next color to play: either PIECE_TYPES.WHITE or PIECE_TYPES.BLACK.
       *
       * @type {String}
       */
      nextPieceType: PIECE_TYPES.BLACK,
    },

    /**
     * Store getters
     *
     * @type {Object}
     * @see https://vuex.vuejs.org/guide/getters.html
     */
    getters: {
      /* eslint-disable jsdoc/require-param, jsdoc/require-example */

      /**
       * The game instance.
       *
       * @returns {Game} The game instance.
       */
      game({ game }) {
        return game;
      },

      /**
       * Current cells, as a flattened array.
       *
       * @returns {Cell[]} A list containing all cells of the board.
       */
      cells({ game, nextPieceType }) {
        return game.board.squares.reduce(
          (cells, row) => [
            ...cells,
            ...row.map(getSquareToCell(game.board, nextPieceType)),
          ],
          [],
        );
      },

      /**
       * Current scores.
       *
       * @returns {Object} An object containing current `white` and `black` scores (numbers).
       */
      scores({ game }) {
        const {
          [PIECE_TYPES.WHITE]: white,
          [PIECE_TYPES.BLACK]: black,
        } = game.board.countByPieceType();
        return { white, black };
      },

      /**
       * Current scores.
       *
       * @returns {Object} An object containing current `white` and `black` statuses:
       * - "idle": do nothing,
       * - "play": player is playing,
       * - "won": player has won.
       */
      statuses({ game, nextPieceType }) {
        if (game.isEnded) {
          const whiteWon = game.getHighScorer() === PIECE_TYPES.WHITE;
          return {
            white: whiteWon ? 'won' : 'idle',
            black: whiteWon ? 'idle' : 'won',
          };
        }

        const nextIsWhite = nextPieceType === PIECE_TYPES.WHITE;
        return {
          white: nextIsWhite ? 'play' : 'idle',
          black: nextIsWhite ? 'idle' : 'play',
        };
      },

      /**
       * Whether the game is in its "start" state.
       *
       * @returns {boolean} `true` if the game is in its "start" state, `false` if a move has been
       * played.
       */
      isAtStart({ game }) {
        return (
          game.board.countByPieceType()[PIECE_TYPES.BLANK] ===
          BOARD_SIZE * BOARD_SIZE - 4
        );
      },

      /* eslint-enable jsdoc/require-param, jsdoc/require-example */
    },

    /**
     * Store mutations.
     *
     * @type {Object}
     */
    mutations: {
      /**
       * Proceed with a game play.
       *
       * @example
       *   // Play on first row, last column
       *   $store.commit('play', { row: 0, col: 7 });
       *
       * @param {Object} state - Previous state.
       * @param {Object} move - Move.
       * @param {number} move.row - Index of row to play.
       * @param {number} move.col - Index of column to play.
       * @returns {undefined} Nothing.
       */
      play(state, { row, col }) {
        const {
          isNextActorPassed,
          pieceType,
          rivalPieceType,
        } = state.game.proceed(row, col);
        state.nextPieceType = isNextActorPassed ? pieceType : rivalPieceType;
      },

      /**
       * Restart the game.
       *
       * @example
       *   // Reatart the game
       *   $store.commit('restart');
       *
       * @param {Object} state - Previous state.
       * @returns {undefined} Nothing.
       */
      restart(state) {
        state.game = new Game();
        state.nextPieceType = PIECE_TYPES.BLACK;
      },
    },
  });

/**
 * The store singleton.
 */
export default createStore();
