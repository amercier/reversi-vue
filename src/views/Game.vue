<template>
  <main class="game">
    <Player
      class="game-player game-player--white"
      color="white"
      :score="$store.getters.scores.white"
      :status="$store.getters.statuses.white"
    />
    <Player
      class="game-player game-player--black"
      color="black"
      :score="$store.getters.scores.black"
      :status="$store.getters.statuses.black"
    />
    <div class="game-content">
      <Board
        class="game-board"
        :cells="$store.getters.cells"
        @moved="handleMove"
      />
    </div>
  </main>
</template>

<script>
import Board from '@/components/Board.vue';
import Player from '@/components/Player.vue';

/**
 * Game page.
 */
export default {
  name: 'Home',
  components: {
    Board,
    Player,
  },
  methods: {
    /**
     * Function called whenever a move is made by a player.
     *
     * @example
     *   // Play on first row, last column
     *   <a v-on:click.prevent="handleMove(0, 7)">
     * @param {number} row - Row index.
     * @param {number} col - Column index.
     * @returns {undefined} Nothing.
     */
    handleMove(row, col) {
      this.$store.commit('play', { row, col });
    },
  },
};
</script>

<style lang="scss">
$player-width: 8rem;
$container-padding: 1.5rem;

.game {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: $container-padding;

  &-player {
    width: $player-width;
    margin: $container-padding;

    &--white {
      order: 0;
    }

    &--black {
      order: 2;
    }
  }

  /**
   * Content layout:
   * 1. Grow as flex child to occupy as much space as possible
   * 2. Constraint to 1:1 ratio by using a combination of width/height + max-width/max-height.
   *    Note: this requires the players to have a fixed width.
   */
  &-content {
    order: 1;

    // All horizontal space, minus players and container padding
    width: calc(100vw - #{2 * $player-width} - #{2 * $container-padding});
    height: calc(100vw - #{2 * $player-width} - #{2 * $container-padding});

    // All vertical space, minus header and container padding
    max-width: calc(100vh - #{$HEADER_HEIGHT} - #{2 * $container-padding});
    max-height: calc(100vh - #{$HEADER_HEIGHT} - #{2 * $container-padding});
  }
}

// Portait mode
@media screen and (max-width: 1024px) and (orientation: portrait) {
  $player-width: 2.5rem;

  .game {
    flex-direction: column;
    padding: $container-padding / 2;

    &-player {
      width: 100%;
      height: $player-width;
    }

    &-content {
      // All horizontal space, minus container padding
      width: calc(100vw - #{$container-padding});
      height: calc(100vw - #{$container-padding});

      // All vertical space, minus header and container padding
      $available-height: calc(
        100vh - #{$HEADER_HEIGHT} - #{2 * $player-width} - #{$container-padding}
      );
      max-width: $available-height;
      max-height: $available-height;
    }

    &-player {
      &--white {
        margin-top: 0;
      }

      &--black {
        margin-bottom: 0;
      }
    }
  }
}
</style>
