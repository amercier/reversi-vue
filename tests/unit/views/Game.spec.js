import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Game from '@/views/Game.vue';
import { createStore } from '@/store'; // TODO Mock

const localVue = createLocalVue();
localVue.use(Vuex);

describe('Game.vue', () => {
  it('renders without crashing', () => {
    const store = createStore();
    expect(() => shallowMount(Game, { store, localVue })).not.toThrow();
  });

  it('renders its components', () => {
    const store = createStore();
    const game = shallowMount(Game, { store, localVue });
    expect(game.element).toMatchSnapshot();
  });
});
