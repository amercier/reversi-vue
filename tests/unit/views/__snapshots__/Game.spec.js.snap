// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game.vue renders its components 1`] = `
<main
  class="game"
>
  <player-stub
    class="game-player game-player--white"
    color="white"
    score="2"
    status="idle"
  />
   
  <player-stub
    class="game-player game-player--black"
    color="black"
    score="2"
    status="play"
  />
   
  <div
    class="game-content"
  >
    <board-stub
      cells="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
      class="game-board"
    />
  </div>
</main>
`;
